### PR TITLE
[MXNET-351] Fix a bug in the MKLDNN integration.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
This is to fix a bug in https://github.com/apache/incubator-mxnet/issues/10580
The bug happens when an NDArray already has a MKLDNN memory with a special MKLDNN format. In some cases (group convolution), the MKLDNN memory uses 5 dimensions, while the NDArray still uses 4 dimensions. In this case, when GetMKLDNNData is called, SetMKLMem will discard the original MKLDNN memory due to incompatible dimensions between MKLDNN memory and NDArray. This messes up the data in the NDArray.

The test code in the PR can actually reproduce the bug in https://github.com/apache/incubator-mxnet/issues/10580

This PR also tries to add C++ unit tests to cover all possible combinations with getting MKLDNN memory from an NDArray.

Thank @TaoLv @pengzhao-intel for finding the root cause of this bug.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

### CODE REVIEW

**Comment by zheng-da** [comment]
@marcoabreu do you know why CI doesn't unpack the right package?
I'm trying to unpack cmake_mkldnn_gpu and it seems the unpacked files don't have libmkldnn.so or any files related to MKLDNN. Does CI cache packages?

```
[ut-cpp-gpu] Running shell script
+ echo Unpacked build/libmxnet.so, build/libmxnet.a, build/3rdparty/dmlc-core/libdmlc.a, build/tests/mxnet_unit_tests, build/3rdparty/openmp/runtime/src/libomp.so from cmake_mkldnn_gpu
Unpacked build/libmxnet.so, build/libmxnet.a, build/3rdparty/dmlc-core/libdmlc.a, build/tests/mxnet_unit_tests, build/3rdparty/openmp/runtime/src/libomp.so from cmake_mkldnn_gpu
+ + echo build/libmxnet.so, build/libmxnet.a, build/3rdparty/dmlc-core/libdmlc.a, build/tests/mxnet_unit_tests, build/3rdparty/openmp/runtime/src/libomp.so
sed -e s/,/ /g
+ xargs md5sum
ec81192c5d42f07d355b5db89c9b343d  build/libmxnet.so
6b09399f30647a8fad23d39c40742576  build/libmxnet.a
d28c0d144be24c657fb71d4dc5a27acf  build/3rdparty/dmlc-core/libdmlc.a
b71eead654d7a783ce34031832f44578  build/tests/mxnet_unit_tests
14167c7484fdd8ddfcb586afdafd5ebd  build/3rdparty/openmp/runtime/src/libomp.so```

I clearly see cmake_mkldnn_gpu has packed libmkldnn.so. Do I miss something?

```
[build-cmake-mkldnn-gpu] Running shell script
+ echo Packing build/libmxnet.so, build/libmxnet.a, build/3rdparty/dmlc-core/libdmlc.a, build/tests/mxnet_unit_tests, build/3rdparty/openmp/runtime/src/libomp.so, build/3rdparty/mkldnn/src/libmkldnn.so, build/3rdparty/mkldnn/src/libmkldnn.so.0 into cmake_mkldnn_gpu
Packing build/libmxnet.so, build/libmxnet.a, build/3rdparty/dmlc-core/libdmlc.a, build/tests/mxnet_unit_tests, build/3rdparty/openmp/runtime/src/libomp.so, build/3rdparty/mkldnn/src/libmkldnn.so, build/3rdparty/mkldnn/src/libmkldnn.so.0 into cmake_mkldnn_gpu
+ echo build/libmxnet.so, build/libmxnet.a, build/3rdparty/dmlc-core/libdmlc.a, build/tests/mxnet_unit_tests, build/3rdparty/openmp/runtime/src/libomp.so, build/3rdparty/mkldnn/src/libmkldnn.so, build/3rdparty/mkldnn/src/libmkldnn.so.0
+ sed -e s/,/ /g
+ xargs md5sum
ec81192c5d42f07d355b5db89c9b343d  build/libmxnet.so
6b09399f30647a8fad23d39c40742576  build/libmxnet.a
d28c0d144be24c657fb71d4dc5a27acf  build/3rdparty/dmlc-core/libdmlc.a
b71eead654d7a783ce34031832f44578  build/tests/mxnet_unit_tests
14167c7484fdd8ddfcb586afdafd5ebd  build/3rdparty/openmp/runtime/src/libomp.so
e9695f70ae19ba9b566b9da82b1d32be  build/3rdparty/mkldnn/src/libmkldnn.so
e9695f70ae19ba9b566b9da82b1d32be  build/3rdparty/mkldnn/src/libmkldnn.so.0
```

**Comment by marcoabreu** [comment]
I think you're missing the ```mx_cmake_mkldnn_lib``` argument in unpack_lib

File: src/ndarray/ndarray.cc [hunk 1]
```diff
@@ -495,24 +495,41 @@ const mkldnn::memory *NDArray::GetMKLDNNDataReorder(
...
       desc2.data.format == GetDefaultFormat(desc2)) {
-    mkldnn_mem_ptr ret(new mkldnn::memory(desc, mem->get_data_handle()));
+    mkldnn_mem_ptr ret(new mkldnn::memory(new_pd, mem->get_data_handle()));
     stream->RegisterMem(ret);
     return ret.get();
-  } else {
-    mkldnn::memory *ret = TmpMemMgr::Get()->Alloc(desc);
+  } else if (same_shape(desc1, desc2)) {
+    // If they have the same shape, we can reorder data directly.
+    mkldnn::memory *ret = TmpMemMgr::Get()->Alloc(new_pd);
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply src/ndarray/ndarray.cc]: ret points to the memory in temp space. it's freed when stream::Submit is called.

File: tests/cpp/operator/mkldnn.cc [hunk 2]
```diff
@@ -28,6 +28,8 @@
 #include "gtest/gtest.h"
 #include "../../src/operator/nn/mkldnn/mkldnn_base-inl.h"
 
+using namespace mxnet;
```
[the remark this answers is not in the document]
Comment (marcoabreu) [reply tests/cpp/operator/mkldnn.cc]: How can I help you here?

[the remark this answers is not in the document]
Comment (zheng-da) [reply tests/cpp/operator/mkldnn.cc]: sorry, i didn't see your comments above. i'll test it first.

**Comment by marcoabreu** [comment]
I agree, this loods strange. But the filehashes of these files look odd as well - could it be possible they're empty or don't exist? Would be good if you could dive a bit deeper there.

**Comment by zheng-da** [comment]
`build/3rdparty/mkldnn/src/libmkldnn.so` and `build/3rdparty/mkldnn/src/libmkldnn.so.0` are essentially the same file. That's why they have the same hash.

**Pull request merged**